rclpy_components/test/test_component_manager.py

```diff
@@ -0,0 +1,198 @@
+        rclpy.spin_until_future_complete(cls.helper_node, future)
+        return future.result()
+
+    def load_node_test(self):
+        load_res = self.__class__.load_node(TEST_COMPOSITION, TEST_COMPOSITION_FOO)
+        assert load_res.success is True
+        assert load_res.error_message == ""
+        assert load_res.unique_id == 1
+        assert load_res.full_node_name == TEST_COMPOSITION_NODE_NAME
+
+        node_name = "renamed_node"
+        node_ns = 'testing_ns'
+        remap_load_res = self.__class__.load_node(
+            TEST_COMPOSITION, TEST_COMPOSITION_FOO, node_name=node_name,
+            node_namespace=node_ns)
+        assert remap_load_res.success is True
+        assert remap_load_res.error_message == ""
+        assert remap_load_res.unique_id == 2
+        assert remap_load_res.full_node_name == f'/{node_ns}/{node_name}'
+
+        list_res: ListNodes.Response = self.__class__.list_nodes()
+        assert len(list_res.unique_ids) == len(list_res.full_node_names) == 2
+
+    def unload_node_test(self):
+        if not self.__class__.unload_cli.wait_for_service(timeout_sec=5.0):
+            raise RuntimeError(f'no unload service found in {self.__class__.container_name}')
+
+        unload_res: UnloadNode.Response = self.__class__.unload_node(1)
+        assert unload_res.success is True
+        assert unload_res.error_message == ""
+
+        # Should be only 1 node left
+        list_res: ListNodes.Response = self.__class__.list_nodes()
+        assert len(list_res.full_node_names) == len(list_res.unique_ids) == 1
+
+        # The index definitely won't exist
+        unload_res: UnloadNode.Response = self.__class__.unload_node(1000)
+        assert unload_res.error_message != ""
+        assert unload_res.success is False
+        list_res: ListNodes.Response = self.__class__.list_nodes()
+        assert len(list_res.full_node_names) == len(list_res.unique_ids) == 1
+
+        # Unload the last node
+        unload_req = UnloadNode.Request()
+        unload_req.unique_id = 2
+        future = self.__class__.unload_cli.call_async(unload_req)
+        rclpy.spin_until_future_complete(self.__class__.helper_node, future)
+        unload_res: UnloadNode.Response = future.result()
+        assert unload_res.success is True
+        assert unload_res.error_message == ""
+
+        # Won't be any node in the container
+        list_res: ListNodes.Response = self.__class__.list_nodes()
+        assert len(list_res.full_node_names) == len(list_res.unique_ids) == 0
+
+    def list_nodes_test(self):
+        container_name = self.__class__.container_name
+        print(f'{container_name}: list_nodes tested within test_load_node and test_unload_node')
+
+    def test_composition_api(self):
+        # Control the order of test
+        self.load_node_test()
+        self.unload_node_test()
+        self.list_nodes_test()
+
+
+class TestComponentManagerMT(TestComponentManager):
```

Comment:
I see a test failure here when I merged this branch into master locally. Any ideay why?

```
============================= test session starts ==============================
platform linux -- Python 3.8.2, pytest-4.6.9, py-1.8.1, pluggy-0.13.0
cachedir: /home/sloretz/bigssd/ros2/build/rclpy_components/.pytest_cache
rootdir: /home/sloretz/bigssd/ros2/src/ros2/rclpy, inifile: pytest.ini
plugins: ament-copyright-0.10.0, ament-pep257-0.10.0, ament-flake8-0.10.0, ament-lint-0.10.0, colcon-core-0.6.0, cov-2.8.1, mock-1.10.4
collecting ...                                 
collected 13 items                                                             

test/test_component_manager.py .F                                        [ 15%]
test/test_component_manager_ut.py ........                               [ 76%]
test/test_copyright.py .                                                 [ 84%]
test/test_flake8.py .                                                    [ 92%]
test/test_pep257.py .                                                    [100%]

=================================== FAILURES ===================================
_________________ TestComponentManagerMT.test_composition_api __________________
test/test_component_manager.py:187: in test_composition_api
    self.load_node_test()
test/test_component_manager.py:130: in load_node_test
    load_res = self.__class__.load_node(TEST_COMPOSITION, TEST_COMPOSITION_FOO)
test/test_component_manager.py:93: in load_node
    raise RuntimeError(f'No load service found in /{cls.container_name}')
E   RuntimeError: No load service found in /TestComponentManagerMT
```

Reply:
I remember to add some components specifically for the test, so the test won't have to rely on any pre-existing components. It seems the testing components are not installed, I'll take a look into this.